aw/Helpers/AWApiHelper.swift

```diff
@@ -0,0 +1,366 @@
+        reach.unit = newReach.unit
+        reach.takeOutLat = newReach.takeOutLat
+        reach.takeOutLon = newReach.takeOutLon
+        reach.rc = newReach.rc
+        reach.delta = newReach.delta
+
+        if let distance = newReach.distanceFrom(
+            location: CLLocation(
+                latitude: DefaultsManager.latitude,
+                longitude: DefaultsManager.longitude)) {
+            reach.distance = distance / 1609
+        }
+        if difficultyRange.contains(1) {
+            reach.difficulty1 = true
+        }
+        if difficultyRange.contains(2) {
+            reach.difficulty2 = true
+        }
+        if difficultyRange.contains(3) {
+            reach.difficulty3 = true
+        }
+        if difficultyRange.contains(4) {
+            reach.difficulty4 = true
+        }
+        if difficultyRange.contains(5) {
+            reach.difficulty5 = true
+        }
+    }
+
+    static func updateRegions(viewContext: NSManagedObjectContext, callback: @escaping UpdateCallback) {
+        let regions = Region.all
+        let dispatchGroup = DispatchGroup()
+        UIApplication.shared.isNetworkActivityIndicatorVisible = true
+        for region in regions {
+            dispatchGroup.enter()
+            fetchReachesByRegion(region: region.code) { (reaches) in
+                let context = NSManagedObjectContext(concurrencyType: .privateQueueConcurrencyType)
+                context.parent = viewContext
+
+                context.perform {
+                    print("aw reaches decoded for \(region.code)")
+                    if let reaches = reaches {
+                        self.createOrUpdateReaches(newReaches: reaches, context: context)
+                    }
+                    do {
+                        try context.save()
+                        print("background context saved")
+                    } catch {
+                        let error = error as NSError
+                        print("unable to save background context \(error) \(error.userInfo)")
+                    }
+                    dispatchGroup.leave()
+                }
+            }
+        }
+        dispatchGroup.notify(queue: .main) {
+            print("update regions complete")
+            viewContext.perform {
+                viewContext.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
+                do {
+                    try viewContext.save()
+                    print("saved view context")
+                } catch {
+                    let error = error as NSError
+                    print("unable to save view context \(error) \(error.userInfo)")
+                }
+            }
+            DefaultsManager.lastUpdated = Date()
+            UIApplication.shared.isNetworkActivityIndicatorVisible = false
+            callback()
+        }
+    }
+
+    static func updateReaches(reachIds: [String], viewContext: NSManagedObjectContext, callback: @escaping UpdateCallback) {
```

Comment:
`viewContext` is sort of a weird name, since the app also has "views"

Reply:
`viewContext` is the same `NSManagedObjectContext` as each view may be using to drive it's FetchedResultController. That way when `updateReaches` runs, it's has child `NSManagedObjectContext`s that it's doing the work in the background on. Once all those child contexts have saved, the `viewContext` merges those changes and the view reflects the changes.